libp2p/messagesHandler.go

```diff
@@ -0,0 +1,668 @@
+		return p2p.ErrTooManyGoroutines
+	}
+
+	handler.throttler.StartProcessing()
+
+	sendable := &SendableData{
+		Buff:  buff,
+		Topic: topic,
+		ID:    handler.idProvider.ID(),
+	}
+	handler.outgoingCLB.GetChannelOrDefault(channel) <- sendable
+	handler.throttler.EndProcessing()
+	return nil
+}
+
+// BroadcastUsingPrivateKey tries to send a byte buffer onto a topic using the topic name as channel
+func (handler *messagesHandler) BroadcastUsingPrivateKey(
+	topic string,
+	buff []byte,
+	pid core.PeerID,
+	skBytes []byte,
+) {
+	handler.BroadcastOnChannelUsingPrivateKey(topic, topic, buff, pid, skBytes)
+}
+
+// BroadcastOnChannelUsingPrivateKey tries to send a byte buffer onto a topic using provided channel
+func (handler *messagesHandler) BroadcastOnChannelUsingPrivateKey(
+	channel string,
+	topic string,
+	buff []byte,
+	pid core.PeerID,
+	skBytes []byte,
+) {
+	go func() {
+		err := handler.broadcastOnChannelBlockingUsingPrivateKey(channel, topic, buff, pid, skBytes)
+		if err != nil {
+			log.Warn("p2p broadcast using private key", "error", err.Error())
+		}
+	}()
+}
+
+// broadcastOnChannelBlockingUsingPrivateKey tries to send a byte buffer onto a topic using provided channel
+// It is a blocking method. It needs to be launched on a go routine
+func (handler *messagesHandler) broadcastOnChannelBlockingUsingPrivateKey(
+	channel string,
+	topic string,
+	buff []byte,
+	pid core.PeerID,
+	skBytes []byte,
+) error {
+	id := peer.ID(pid)
+	sk, err := libp2pCrypto.UnmarshalSecp256k1PrivateKey(skBytes)
+	if err != nil {
+		return err
+	}
+
+	err = handler.checkSendableData(buff)
+	if err != nil {
+		return err
+	}
+
+	if !handler.throttler.CanProcess() {
+		return p2p.ErrTooManyGoroutines
+	}
+
+	handler.throttler.StartProcessing()
+
+	sendable := &SendableData{
+		Buff:  buff,
+		Topic: topic,
+		Sk:    sk,
+		ID:    id,
+	}
+	handler.outgoingCLB.GetChannelOrDefault(channel) <- sendable
+	handler.throttler.EndProcessing()
+	return nil
+}
+
+func (handler *messagesHandler) checkSendableData(buff []byte) error {
+	if len(buff) > maxSendBuffSize {
+		return fmt.Errorf("%w, to be sent: %d, maximum: %d", p2p.ErrMessageTooLarge, len(buff), maxSendBuffSize)
+	}
+	if len(buff) == 0 {
+		return p2p.ErrEmptyBufferToSend
+	}
+
+	return nil
+}
+
+// RegisterMessageProcessor registers a message process on a topic. The function allows registering multiple handlers
+// on a topic. Each handler should be associated with a new identifier on the same topic. Using same identifier on different
+// topics is allowed. The order of handler calling on a particular topic is not deterministic.
+func (handler *messagesHandler) RegisterMessageProcessor(topic string, identifier string, msgProcessor p2p.MessageProcessor) error {
+	if check.IfNil(msgProcessor) {
+		return fmt.Errorf("%w when calling messagesHandler.RegisterMessageProcessor for topic %s",
+			p2p.ErrNilValidator, topic)
+	}
+
+	topicProcs := handler.topicsHandler.GetTopicProcessors(topic)
```

Comment:
I do not think this is equivalent with the old code:
```GO
netMes.mutTopics.Lock()
defer netMes.mutTopics.Unlock()
topicProcs := netMes.processors[topic]
if topicProcs == nil {
	topicProcs = newTopicProcessors()
	netMes.processors[topic] = topicProcs

	err := netMes.pb.RegisterTopicValidator(topic, netMes.pubsubCallback(topicProcs, topic))
	if err != nil {
		return err
	}
}

err := topicProcs.addTopicProcessor(identifier, handler)
if err != nil {
	return fmt.Errorf("%w, topic %s", err, topic)
}
```

because the previous version was contained all calls as a critical section
Suggestion to drop the usage of the topicsHandler

Reply:
indeed. dropped the component and reverted the code as before